### ISSUE TITLE
Cyclic references when using bidirectional many-to-one connection

### ISSUE THREAD

**Issue by denisvmedia** [cayleygraph/cayley#752]
```
package main

import (
	"context"
	"fmt"
	"github.com/cayleygraph/cayley"
	"github.com/cayleygraph/cayley/graph"
	_ "github.com/cayleygraph/cayley/graph/kv/bolt"
	"github.com/cayleygraph/cayley/quad"
	"github.com/cayleygraph/cayley/schema"
	"github.com/cayleygraph/cayley/voc"
	"io/ioutil"
	"log"
	"os"

	// Import RDF vocabulary definitions to be able to expand IRIs like rdf:label.
	_ "github.com/cayleygraph/cayley/voc/core"
)

type Cluster struct {
	_     struct{} `quad:"@type > gc:Cluster"`
	ID    quad.IRI `json:"@id"`     // tag @id is a special one - graph node value will be stored in this field
	Name  string   `json:"gc:name"` // field name (predicate) may be written as json field name
	Hosts []Host   `json:"-" quad:"gc:cluster < *,optional"`
	Nodes []Node   `json:"-" quad:"gc:cluster < *,optional"`
}

type Host struct {
	_       struct{} `quad:"@type > gc:Host"`
	ID      quad.IRI `json:"@id"`     // tag @id is a special one - graph node value will be stored in this field
	Name    string   `json:"gc:name"` // field name (predicate) may be written as json field name
	Cluster *Cluster `json:"-" quad:"gc:cluster > *"`
	Node    *Node    `json:"-" quad:"gc:host < *,optional"`
}

type Node struct {
	_       struct{} `quad:"@type > gc:Node"`
	ID      quad.IRI `json:"@id"`     // tag @id is a special one - graph node value will be stored in this field
	Name    string   `json:"gc:name"` // field name (predicate) may be written as json field name
	Cluster *Cluster `json:"-" quad:"gc:cluster"`
	Host    *Host    `json:"gc:host"`
}

func checkErr(err error) {
	if err != nil {
		log.Fatal(err)
	}
}

func main() {
	voc.RegisterPrefix("gc:", "http://gc.gc/")
	sch := schema.NewConfig()

	tmpdir, err := ioutil.TempDir("", "example")
	checkErr(err)

	defer os.RemoveAll(tmpdir) // clean up

	// Initialize the database
	err = graph.InitQuadStore("bolt", tmpdir, nil)
	checkErr(err)

	// Open and use the database
	store, err := cayley.NewGraph("bolt", tmpdir, nil)
	checkErr(err)
	defer store.Close()
	qw := graph.NewWriter(store)

	// Save an object
	cl := Cluster{
		ID:   quad.IRI("gc:cluster1").Full().Short(),
		Name: "Cluster1",
	}
	fmt.Printf("saving: %+v\n", cl)
	id, err := sch.WriteAsQuads(qw, cl)
	checkErr(err)
	err = qw.Flush()
	checkErr(err)
	fmt.Println("id for object:", id, "=", cl.ID) // should be equal

	// Save an object
	h := Host{
		ID:      quad.IRI("gc:host1").Full().Short(),
		Name:    "Host1",
		Cluster: &cl,
	}
	fmt.Printf("saving: %+v\n", h)
	id, err = sch.WriteAsQuads(qw, h)
	checkErr(err)
	err = qw.Flush()
	checkErr(err)
	fmt.Println("id for object:", id, "=", h.ID) // should be equal

	// Save an object
	h = Host{
		ID:      quad.IRI("gc:host2").Full().Short(),
		Name:    "Host2",
		Cluster: &cl,
	}
	fmt.Printf("saving: %+v\n", h)
	id, err = sch.WriteAsQuads(qw, h)
	checkErr(err)
	err = qw.Flush()
	checkErr(err)
	fmt.Println("id for object:", id, "=", h.ID) // should be equal

	// Save an object
	n := Node{
		ID:      quad.IRI("gc:node1").Full().Short(),
		Name:    "Node1",
		Host:    &h,
		Cluster: &cl,
	}
	fmt.Printf("saving: %+v\n", n)
	id, err = sch.WriteAsQuads(qw, n)
	checkErr(err)
	err = qw.Flush()
	checkErr(err)
	fmt.Println("id for object:", id, "=", n.ID) // should be equal

	// Get object by id
	var somehost Host
	err = sch.LoadToDepth(nil, store, &somehost, 0, quad.IRI("gc:host1"))
	checkErr(err)
	fmt.Printf("loaded: %+v\n", somehost)

	// Get object by id
	var somenode Node
	err = sch.LoadToDepth(nil, store, &somenode, 0, quad.IRI("gc:node1"))
	checkErr(err)
	fmt.Printf("loaded: %+v\n", somenode)

	// Or get all objects of type Cluster
	var clusters []Cluster
	err = sch.LoadToDepth(nil, store, &clusters, 0)
	checkErr(err)
	fmt.Printf("clusters: %+v\n", clusters)
	for _, v := range clusters {
		fmt.Println(v.Name)
	}

	fmt.Println()

	// Print quads
	fmt.Println("\nquads:")
	ctx := context.TODO()
	it := store.QuadsAllIterator()
	for it.Next(ctx) {
		fmt.Println(store.Quad(it.Result()))
	}
}
```

And it works fine, apart that I don't have proper entities in Cluster.Nodes and Cluster.Hosts. Their data are not set (only ids are available). This is of because I have depth set to `0`. But if I set it to `-1`, the program crashes: `panic: reflect.Set: value of type *main.Host is not assignable to type main.Host`.

This is not the case when it comes to one-to-one links. But t seems that bi-directional many-to-one links []Type <-> *Type are not possible. Is it a bug, a feature or a known issue?

**Comment by dennwc** [comment]
It's definitely a bug that it panics and doesn't reuse the node for depth = 0.

But the definition that you are using is also invalid. If you have loops, the library will try to reuse node pointers, and in case of `[]Node` it won't be able to assign an existing pointer value without copying it, which may lead to weird bugs. Try changing the type to `[]*Node`, maybe it will solve an issue for you.

**Comment by denisvmedia** [comment]
I tried both: `[]Node` and `[]*Node`. The problem was the same. In fact, my initial definition was `[]*Node`.

**Comment by dennwc** [comment]
Should be fixed in 7b6e2662ddff019dd535dbb5573ca74c9b17c4b1. Feel free to reopen if it's not.
